mava/systems/sac/anakin/ff_hasac.py

```diff
@@ -0,0 +1,727 @@
+        next_obs=obs,
+    )
+
+    rb = fbx.make_item_buffer(
+        max_length=int(cfg.system.buffer_size),
+        min_length=int(cfg.system.explore_steps),
+        sample_batch_size=int(cfg.system.batch_size),
+        add_batches=True,
+    )
+    buffer_state = replicate(rb.init(init_transition))
+
+    networks = (actor_network, q_network)
+    optims = (actor_opt, q_opt, alpha_opt)
+
+    # Reset env.
+    n_keys = cfg.arch.num_envs * cfg.arch.n_devices * cfg.system.update_batch_size
+    key_shape = (cfg.arch.n_devices, cfg.system.update_batch_size, cfg.arch.num_envs, -1)
+    key, reset_key = jax.random.split(key)
+    reset_keys = jax.random.split(reset_key, n_keys)
+    reset_keys = jnp.reshape(reset_keys, key_shape)
+
+    # Keys passed to learner
+    first_keys = jax.random.split(key, (cfg.arch.n_devices * cfg.system.update_batch_size))
+    first_keys = first_keys.reshape((cfg.arch.n_devices, cfg.system.update_batch_size, -1))
+
+    env_state, first_timestep = jax.pmap(  # devices
+        jax.vmap(  # update_batch_size
+            jax.vmap(env.reset),  # num_envs
+            axis_name="batch",
+        ),
+        axis_name="device",
+    )(reset_keys)
+    first_obs = first_timestep.observation
+
+    t = jnp.zeros((cfg.arch.n_devices, cfg.system.update_batch_size), dtype=int)
+
+    # Initial learner state.
+    learner_state = LearnerState(
+        first_obs, env_state, buffer_state, params, opt_states, t, first_keys
+    )
+    return (env, eval_env), networks, optims, rb, learner_state, target_entropy, logger, key
+
+
+def make_update_fns(
+    cfg: DictConfig,
+    env: MarlEnv,
+    networks: Networks,
+    optims: Optimisers,
+    rb: TrajectoryBuffer,
+    target_entropy: chex.Array,
+) -> Tuple[
+    Callable[[LearnerState], Tuple[LearnerState, Metrics]],
+    Callable[[LearnerState], Tuple[LearnerState, Tuple[Metrics, Metrics]]],
+]:
+    """Create the update functions for the learner.
+
+    Args:
+    ----
+        cfg: System configuration.
+        env: The environment.
+        networks: Tuple of actor and critic networks.
+        optims: Tuple of actor, critic and alpha optimisers.
+        rb: The replay buffer.
+        target_entropy: The target entropy.
+
+    Returns:
+    -------
+        Tuple of (explore_fn, update_fn).
+        Explore function is used for initial exploration with random actions.
+        Update function is the main learning function, it both acts and learns.
+    """
+    actor_net, q_net = networks
+    actor_opt, q_opt, alpha_opt = optims
+
+    full_action_shape = (cfg.arch.num_envs, *env.action_spec().shape)
+
+    # losses:
+    def q_loss_fn(
+        q_params: QVals, obs: Array, action: Array, target: Array
+    ) -> Tuple[Array, Metrics]:
+        q1_params, q2_params = q_params
+        # Concat all actions and tile them for num agents to create joint actions for all agents
+        joint_action = get_joint_action(action)  # (B, A, Act) -> (B, A, A * Act)
```

Comment:
Just a note to please update all the shape comments. Can we please do this everywhere? 
```suggestion
        joint_action = get_joint_action(action)  # (B, N, A) -> (N, N, N * A)
```